[Wnayar] ip

Wnayar opened the pull request:
# Bill 
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/l))

Bill frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is, 
1. download it from [here](https://github.com/Wnayar/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing todo task
- [x] Managing deadline task
- [x] Managing event task
- [x] Listing tasks
- [x] Filtering tasks

If you are a Java programmer, you can use it to practice Java too. Here's  the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Wnayar committed:
New chatbot name with initial skeletal version

Wnayar committed:
Chatbot echo commands entered by the user

Wnayar committed:
Add list functionality to store and show user commands

Wnayar committed:
Add tasks class with ability to mark as done or not done

Wnayar committed:
Refactor code structure to incorporate OOP concepts

Wnayar committed:
Add support for tracking three types of tasks, todo, deadline and event

Wnayar committed:
Partial automate text UI test

Wnayar committed:
update batch script for future automated text UI test

Wnayar committed:
Handle errors with custom BillException class

Wnayar committed:
Add support for deleting tasks from the list

Wnayar committed:
Use java enums as router selection

Wnayar committed:
Save tasks in the hard disk when the list changes

Wnayar committed:
Merge branch 'branch-Level-7'

Wnayar committed:
Teach the chatbot how to understand dates and times for deadline and display it

Wnayar committed:
Merge branch 'branch-Level-8'

Wnayar committed:
Refactor code to be more OOP with new UI, Storage, Parser and Tasklist classes

Wnayar committed:
Put all classes in one package called bill

Wnayar committed:
Merge branch 'branch-A-Packages'

Wnayar committed:
Merge branch 'add-gradle-support'

Wnayar committed:
Add gradle and confirguration to automate project builds

Wnayar committed:
Add Junit tests for some deadline and task functions

Wnayar committed:
Configure build.gradle file to set JAR file output name as bill when run task shadowJar in gradle

Wnayar committed:
Add javadocs to Bill class

Wnayar committed:
Partion and space import statements by groups

Wnayar committed:
Add empty template for find a task method

Wnayar committed:
Add find a task functionality to search for tasks with a keyword which matches the task description

Wnayar committed:
Add JavaDocs to at least half of the non-private classes and methods

Wnayar committed:
Merge branch 'branch-A-JavaDoc'

Wnayar committed:
Tweak Code to comply with CS2103T java coding standards

Wnayar committed:
Merge branch 'branch-A-CodingStandard' to main and resolves merge conflicts

Wnayar committed:
Merge branch 'branch-Level-9' into main, resolve merge conflict

foothaddeus commented on src/main/java/bill/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+        this.by = parseUserDateHelper(by);
+    }
+
+    private LocalDate parseUserDateHelper(String dateStr) {
+        // Allow two types of inputs, second format is what is from Bill.txt, user can use that format too for cml
+        // Can improve this, ideally don't use try catch as control flow, could also try just allowing one input
+        DateTimeFormatter formatterUserInput = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        DateTimeFormatter formatterFileInput = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        try {
+            return LocalDate.parse(dateStr, formatterUserInput);
+        } catch (DateTimeParseException ex) {
+            return LocalDate.parse(dateStr, formatterFileInput);
+        }
+    }
+
+    private String dateStringRepresentation() {
+        return by.format(DateTimeFormatter.ofPattern("MMM dd yyy"));
+    }
```
Would it be better to name this method "parseUserDate" instead? That way it is not so long and fits the coding standard of naming methods as verbs better.

Wnayar replied:
Yep I agree, functions should be verbs. Will change accordingly.

foothaddeus commented on src/main/java/bill/Event.java:
```diff
@@ -0,0 +1,29 @@
+package bill;
+
+/**
+ * The Event class extends the Task class, and allows creating tasks with a range of time,
+ * from a certain start period to a certain end period.
+ */
```
I like how detailed you are for your javadoc comments!

foothaddeus commented on src/main/java/bill/Ui.java:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+    public void silentLoadingData(Storage storage, ArrayList<Task> userList, TaskList tasks) {
+        // load data in, but silence print messages, as leveraging custom functions which have print statements
+        PrintStream originalOutput = System.out;
+        // temp output to hide print messages, so can leverage previously built methods which have print statements
+        OutputStream silence = new OutputStream() {
+            @Override
+            public void write(int b) throws IOException {
+                //do nothing
+            }
+        };
+        try {
+            System.setOut(new PrintStream(silence));
+            storage.loadStorage(this, userList, tasks);
+        } catch (IOException ex) {
+            System.setOut(originalOutput);
+            System.out.println(ex.getMessage());
+            return;
+        } catch (BillException ex) {
+            System.setOut(originalOutput);
+            System.out.println("There is a formatting issue with the bill.txt file, it is related to:");
+            System.out.println(ex.getMessage());
+            return;
+        }
+        System.setOut(originalOutput);
+    }
```
Perhaps this could be named "silenceLoadingData" instead?

Wnayar replied:
I agree, the function should be verb. Will change accordingly.

foothaddeus commented on src/main/java/bill/Parser.java:
```diff
@@ -0,0 +1,126 @@
+        return userCommand.split(" ");
+    }
+
+    public String handleFindParser(String[] parsedInput) throws BillException {
+        // data validation
+        if (parsedInput.length < 2) {
+            throw new BillException("Please provide a second argument for the find command, such as a keyword,"
+                        + " the format for find is: find <>, where <> is a single keyword");
+        }
+        if (parsedInput.length > 2) {
+            throw new BillException("Please ensure the following format for the find command:"
+                        + " find <>, where <> is a single keyword");
+        }
+        return parsedInput[1];
+    }
```
Although not necessary, I think it would be good to have javadoc comments for more of your public methods, especially those like this which are not as straightforward to understand!

Wnayar replied:
I agree. Will add in the future.  Was conforming to the minimum requirement of >= 50% of javadocs for public classes and methods.

foothaddeus reviewed (commented):
Overall, looks good!

Wnayar committed:
Conform to CS2103T coding standardsand rename functions to verbs

Wnayar committed:
Allow use of gradle to conduct style checks

Wnayar committed:
Merge branch 'branch-A-CheckStyle'

Wnayar committed:
Add javadocs to all public methods and conform to CS2103T style check

Wnayar committed:
Refactor code to make router in ui return strings

Wnayar committed:
Add a Gui to the chatbot with Javafx technology implementation

Wnayar committed:
Merge branch 'branch-Level-10'

Wnayar committed:
Update the pictures of the user and bill in the Gui

ngjianyi commented on src/main/java/bill/Storage.java:
```diff
@@ -0,0 +1,128 @@
+     * @throws IOException If there is an error reading from the bill.txt file.
+     * @throws BillException If there is an unrecognisable format in the bill.txt file.
+     */
+    public void loadStorage(Ui ui, ArrayList<Task> userList, TaskList tasks) throws IOException, BillException {
```
I like how the function is named after the action!

ngjianyi commented on src/main/java/bill/TaskList.java:
```diff
@@ -0,0 +1,115 @@
+     * @param targetTaskNumber The index of the targeted task from the userList to mark or unmark it.
+     * @param markOrUnmark Identifier of mark or unmark, which determines if the target task will be marked or unmarked.
+     */
+    public String markOrUnmarkTask(int targetTaskNumber, String markOrUnmark, Storage storage) throws IOException {
```
I like the clear naming of functions! Perhaps it may be better to split this function into two separate functions?

ngjianyi commented on src/main/java/bill/Ui.java:
```diff
@@ -0,0 +1,261 @@
+     * @param storage All helper methods associated with storing tasks
+     * @param tasks ALl helper methods associate with manipulating and showing tasks
+     */
+    public void handleUserCommands(Storage storage, TaskList tasks) {
```
Perhaps this function can be named to specify the exact task it is doing?

ngjianyi commented on src/main/java/bill/Task.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Task {
+    private String description;
+    private boolean isDone;
```
Good use of naming the variable to sound like a boolean!

Wnayar committed:
Implement Java Varags feature to handleFindParser method in the Parser class

Wnayar committed:
Merge branch 'branch-A-Varargs'

Wnayar committed:
Update the storage path

The path is absolute and uses desktop.

The absolute path will not run across different users.

Let's use a relative path to where the file is located.

Using relative pathing is preferable to allow the file to be run
at different locations, and not just at the desktop.

Wnayar committed:
Add assertion statements

The codebase does not have assertion statements which is important
to ensure correct state of program.

Let's add assertion statements to the appropriate places.

Using assertion statements helps to catch programmer related
mistakes based on the expected state of the program.

Wnayar committed:
Merge pull request #1 from Wnayar/branch-A-Assertions

Add assertion statements

Wnayar committed:
Merge branch 'master' of https://github.com/Wnayar/ip

* 'master' of https://github.com/Wnayar/ip:
  Add assertion statements

Wnayar committed:
Merge branch 'master' into branch-A-CodeQuality

Wnayar committed:
Merge branch 'master' into branch-A-Streams

Wnayar committed:
Improve code quality

Most classes do not follow SLAP, naming and length conventions.

Let's implement SLAP, ensure functions are named as verbs
and variables as nouns and ensure functions are below 30 lines of
code.

Implementing good code quality improves readability and
understandability which allows for better collaboration and
maintainability.

Wnayar committed:
Merge pull request #2 from Wnayar/branch-A-CodeQuality

Branch a code quality

Wnayar committed:
Merge branch 'master' into branch-A-Streams

Wnayar committed:
Use a stream to filter the list in the find function

Wnayar committed:
Merge pull request #3 from Wnayar/branch-A-Streams

Branch a streams

Wnayar committed:
Add Github Actions to set up Continuous Integration (CI)

Wnayar committed:
Merge branch 'branch-A-CI'

Wnayar committed:
Add a BCD-Extension to add a help command

Wnayar committed:
Merge branch 'branch-BCD-Extension'

Wnayar committed:
Fix help command and add javadoc to it

Wnayar committed:
Improve javadoc for loadData method

Wnayar committed:
Ensure Bill Bot prevents user input after bye command

Wnayar committed:
Add a Guide class for the user command help

Wnayar committed:
Add more Junit tests

Wnayar committed:
Add a Ui image of Bill bot

Wnayar committed:
Add a Bill user guide

Wnayar committed:
Merge branch 'branch-A-UserGuide'

Wnayar committed:
Update readme

Wnayar committed:
Remove irrelevant comments
